### ISSUE TITLE
Timings discussion

### ISSUE THREAD

**Issue by VoidAngel** [TheBusyBiscuit/Slimefun4#582]
## **Description**
Anyone with a moderately sized server dependent on Slimefun knows that the cargo system can cause a disproportionately large amount of lag. I'm opening this in hopes to have a discussion with the Slimefun community on this issue, and ultimately find ways to improve the cargo system. This may also be helpful for others that do not currently know their options.

Options we have so far:
1) Disable cargo.
2) Disable advanced output nodes (These nodes seem to cause the heavy methods to be called when not needed, is there a way to stop this? & I could be wrong on this)
3) Change custom-ticker-delay in the config to a higher value. This means cargo related tasks will happen less often. Its a decent compromise.
4) Find points that can be optimized in the current code.
5) Recode the entire cargo system.

Does anyone have any ideas? extra info? other possible options?

## **Evidence / Proof**
Lag spikes every few seconds:
https://pastebin.com/zptFAg5K

## **Environment**
Paper spigot 1.12.1 1170
CS-Corelib 1.5.16
Slimefun 4.1.13

**Comment by AtomicScience** [comment]
Well, does any cargo-related stuffs run asynchronously?

**Comment by MasterGabeMOD** [comment]
I agree, something needs to be done with the cargo system. While I have seen some improvement in V4.1.13, it's not enough. I don't have a full understanding how the cargo system works as I've never played with it. My guess is they behave similar to hoppers with items being ported between several nodes. 

As VoidAngel mentioned above, I believe if a tick delay was put in the config for us to delay the cargo tasks, it may resolve these issues.

**Comment by VoidAngel** [comment]
Lucky you @MasterGabeMOD, its already there xD its called custom-ticker-delay. 12 is default I believe. I've changed mine to 24 which helps a lot. Im happy with how the server runs and how the cargo performs with the value being 24. Sadly, I know that players will always be trying to push the limits of what they can do and in the end they will just make more nodes to counter the loss in speed. Because of that I'd say its more of a "band-aid" than a solution.

@AtomicScience I dont believe any of it is async. Correct me if I'm wrong, but wouldn't that cause concurrent modification exceptions?

**Comment by TheBusyBiscuit** [comment]
This discussion pops up again and again.
It is very easy to shout "It must be improved" as a Server Owner.

We know it is slow but lag is not (always) the result of bad programming...
Lag just means that the Server has to do too much in a too short time period.

To go over @VoidAngel 's "options":

1. You can do that.
2. Not that I am aware of
3. A bandaid as stated before.
4. Find a needle in a pile of hay, Im talking about a pile that is roughly a million lines of code. And even then the improvement is at most minor .
5. Nobody got time for that. And it would be a huge huge rewrite that may not solve any problems.

And no, it is not async, like @VoidAngel stated, it would throw errors if it was.
"Make it async" is never a brilliant solution.
Bukkit is not multi threaded, making it hard for plugins to use multiple threads.
Anything that has to interact with Minecraft/Bukkit is forced to be sync.

**Comment by AtomicScience** [comment]
Well, TheBusyBuscuit is right again: nobody have time and aspiration to rewrite cargo system, and, probably, it won't help. My server timings show, what the whole Slimefun isn't cause of any lags, probably your problem is unique, or you just do something wrong

**Comment by VoidAngel** [comment]
I appreciate the feedback <3
And yeah, I don't think a cargo recode will happen either. I would do it if I had the knowledge, but sadly I'm not that advanced in programming. However, I may attempt to lessen the rapid heavy method calls that advanced output nodes cause. I have no idea if its even possible at the moment, or if it will even help, so I'll have to do some fiddling.
Also @AtomicScience, its just the way my server is set up. Players are required to use Slimefun in order to play at all. Many of them become experts and end up with hundreds of machines and large cargo networks to connect them all. Its kind of funny, when the most experienced players are online the server struggles and TPS dips. I sorta feel bad for those people D:

**Comment by John000708** [comment]
Some sort of node limit per player could be added. I am not sure how practical that would be and there are plenty of ways to bypass it with alt accounts.

**Comment by VoidAngel** [comment]
That's actually a pretty good idea. I didn't think about that. And yes, some people would bypass with alts but I think the majority wouldn't know how, or would just be too lazy xD So it would lessen the problem to some degree. It would also promote players being more efficient with the nodes that they do have.

**Comment by AtomicScience** [comment]
Well, @VoidAngel , admit, what industrial survival at vanilia, even with such amazing plugin as Slimefun doesn't equals industrial survival with mods: it's laggy, it's not such pretty as mods, and... of course, these networks and other Slimefun stuffs overload your project a lot, so you have to run server at very powerful machine or limit your players in Slimefun-related possibilities or. So, I think, it's almost impossible to improve perfomance of this plugin becouse of limitations of Bukkit platform.

**Comment by VoidAngel** [comment]
I understand. I don't blame anyone, and I'm not going to scream and kick and demand someone fix my problems for me xD People already do that to me enough. I have considered the possibility that optimization simply cannot be done due to limits with minecraft, bukkit, java, my own machine, or what ever else. However, don't think I'm going to give up without at least trying ;D I've tried on my own for a while, but the other day I decided that maybe I should open a discussion just to see what other people think on the subject. Thousands of people love this plugin and want the best for it, including me. 

John's response has given me something to work with, and I may try it on my own version. I don't want to limit the majority of players, but Ill be honest... A select few of the players on my server have created systems that are almost ridiculous xD
